### PR TITLE
[improve][client] Add retry when readTailMessages fails

### PR DESCRIPTION
**Pull request by heesung-sn**
### Motivation

If TableView is interrupted while readTailMessages, it will soon have outdated data as it stops reading more messages. Instead, in this case, TableView needs to retry.

### Modifications

Changed readTailMessages to run indefinitely even if it fails to read.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:
- Added a unit test.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: skipped as this is a simple change.

### CODE REVIEW

**Comment by lifepuzzlefun** [comment]
I understand the idea of this patch. but from a api user's view. it is confusing that what is an `interrupted`, and how should the api user react to the interrupted. I think should choose how to expose the internal exception on TableView to the api user. or we handle the exception inside the client with some retry method. : - )

**Comment by heesung-sn** [comment]
1. retry forever upon failures.
2. throw RuntimeException from the data accessor apis, if already interrupted.(get, entrySet, keySet)
3. expose table view state(e.g isInterrupted)

It appears that we have the above options.

If tableview is supposed to guarantee the liveness, I think we can go to the option 1 instead.

**Comment by heesung-sn** [comment]
For the option 3, by this pr, callers could create new tableviews, by constantly monitoring if the table view is interrupted or not.

**Comment by heesung-sn** [comment]
To expand this discussion a little further, I think the current tableview works under best-effort delivery and needs some improvement in the following areas for mission-critical data replication in production. 


- guarantee liveness(better error handling. e.g infinite retry)
- handle slowness (how to handle slow consumer, e.g self-kill if too slow. If the consumer barely receives messages due to bugs, network, and etc.)
- expose operational metrics(table view state(retrying, active), consume latency, highest messageId(highest logical sequence number))

Currently, it is hard or impossible for callers to check them outside of tableview. And this pr is trying to expose the worst-case tableview state,  where tableview is interrupted and stopped to consume messages. We could change the func name to isAlive, isConnected, isClosed, isStopped, or other. I found isInterrupted is a good choice here, as it tells if tableview is "interrupted" by some unknown reason.

**Comment by heesung-sn** [comment]
I updated the code to retry upon failure. @lifepuzzlefun

**Comment by heesung-sn** [comment]
@lifepuzzlefun I updated the code to "retry-forever" upon `readTailMessages` failure. I assume the current "fail-stop" is a bug.

**Comment by heesung-sn** [comment]
@lifepuzzlefun I updated the code to "retry-forever" upon `readTailMessages` failure. I assume the current "fail-stop" is not expected. 

Another thing to note is that since the reader acks messages immediately, this "retry" will not retry the handling logic with the same message. (It will move on and retry the next message)

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java [hunk 1]
```diff
@@ -266,18 +270,15 @@ private void readTailMessages(Reader<T> reader) {
...
                     readTailMessages(reader);
                 }).exceptionally(ex -> {
-                    logException(
-                            String.format("Reader %s was interrupted while reading tail messages.",
-                                    reader.getTopic()), ex);
+                    if (ex.getCause() instanceof PulsarClientException.AlreadyClosedException) {
+                        log.error("Reader {} was closed while reading tail messages.",
+                                reader.getTopic(), ex);
+                    } else {
+                        log.warn("Reader {} was interrupted while reading tail messages. "
```
Comment (Demogorgon314) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java:277]: Do we need to delay few milliseconds before retry?

Comment (heesung-sn) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java:277]: I think we can rely on consumer.receive logic for waiting. I don't think we need additional delay here.

**Review by Demogorgon314 — APPROVED** [review]

**Review by lifepuzzlefun — APPROVED** [review]

**Review by merlimat — APPROVED** [review]

**Pull request merged**